504/adr_IAsyncEnumerable_IActionResult (#137)
microsoft/carbon-aware-sdk: docs/architecture/decisions/0005-IAsyncEnumerable-HttpResponseExceptionFilter.md

# 5. WebAPI to use IAsyncEnumerable to stream potentially large responses.

## Status
Denied

## Context

> There are currently two endpoints (`emissions/forecasts/batch`, `emissions/average-carbon-intensity/batch`) that require passing a payload as an array of items. This array can have as many items as the user wants, and it can take sometime to be processed which incurs in a delay to the client, which might feel that the service is irresponsive. Also, since these requests return enumerable items, and those are buffered before the client gets it, it impacts the overall memory footprint of the WebApp. Changing the signature of these endpoints to return an `IAsyncEnumerable` collection streams the response and helps to deal with these memory concerns.

Reviewer: Nit:  'sometime' should be 'some time' and 'incurs in' should be 'incurs' (or even better would be the word 'creates').

> # 5. Using IAsyncEnumerable partial failure won't be handled by HttpResponseExceptionFilter

Reviewer 2: Let's reword this as a decision to make:

```markdown
# 5. WebAPI to use IAsyncEnumerable to stream potentially large responses

## Status

Denied
```

OR the positive version::
 
```markdown
# 5. WebAPI to use IActionResult for potentially large resonses.

## Status

Accepted
```

Author: Good. I'll add the former.

## Status
Accepted

## Context

> There are currently two endpoints (`emissions/forecasts/batch` , `emissions/average-carbon-intensity/batch`) that require passing a payload as an array of items. This array can be have as many items as the user wants, and it can take sometime to be processed which incurs in a delay to the client, which might feel that the service is irresponsive. Also, since these requests return enumerable items, and those are buffered before the client gets it, it impacts the overall memory footprint of the WebApp. By changing the signature of these endpoints to return an `IAsyncEnumerable` collection, it would help to deal with these issues.

Reviewer 2: ```suggestion
There are currently two endpoints (`emissions/forecasts/batch` , `emissions/average-carbon-intensity/batch`) that require passing a payload as an array of items. This array can be have as many items as the user wants, and it can take sometime to be processed which incurs in a delay to the client, which might feel that the service is irresponsive. Also, since these requests return enumerable items, and those are buffered before the client gets it, it impacts the overall memory footprint of the WebApp. Changing the signature of these endpoints to return an `IAsyncEnumerable` collection streams the response and helps to deal with these memory concerns.
```

> ## Decision

Reviewer 2: Depending on which approach you take, `Use IAsyncEnumerable: Denied` or `Use IActionResult: Accepted`, can we put the relevant pros into the  `## Decision`, and the cons into the `## Consequences` sections.
